geo/src/algorithm/intersects/line_string.rs

```diff
@@ -3,31 +3,141 @@ use crate::BoundingRect;
+symmetric_intersects_impl!(LineString<T>, MultiLineString<T>);
+
+impl<T> Intersects<Polygon<T>> for LineString<T>
+where
+    T: GeoNum,
+    Line<T>: Intersects<Line<T>>,
+    Coord<T>: Intersects<Polygon<T>>,
+{
+    fn intersects(&self, rhs: &Polygon<T>) -> bool {
+        if self.is_empty() || rhs.is_empty() {
+            return false;
+        }
+        if has_disjoint_bboxes(self, rhs) {
+            return false;
+        }
+        // if no lines intersections, then linestring is either disjoint or within the polygon
+        self.0[0].intersects(rhs)
+            || self
+                .lines()
+                .any(|l| rhs.lines_iter().any(|other| l.intersects(&other)))
+    }
+}
+
+impl<T> Intersects<MultiPolygon<T>> for LineString<T>
 where
-    T: CoordNum,
-    Line<T>: Intersects<G>,
-    G: BoundingRect<T>,
+    T: GeoNum,
+    Line<T>: Intersects<Line<T>>,
+    Coord<T>: Intersects<Rect<T>>,
 {
-    fn intersects(&self, geom: &G) -> bool {
-        if has_disjoint_bboxes(self, geom) {
+    fn intersects(&self, rhs: &MultiPolygon<T>) -> bool {
+        if has_disjoint_bboxes(self, rhs) {
             return false;
         }
-        self.lines().any(|l| l.intersects(geom))
+        // splitting into `LineString intersects Polygon`
+        rhs.iter().any(|poly| self.intersects(poly))
     }
 }
 
-// Blanket implementation from LineString<T>
-impl<T, G> Intersects<G> for MultiLineString<T>
+impl<T> Intersects<Rect<T>> for LineString<T>
 where
-    T: CoordNum,
-    LineString<T>: Intersects<G>,
-    G: BoundingRect<T>,
+    T: GeoNum,
+    Line<T>: Intersects<Line<T>>,
+    Coord<T>: Intersects<Rect<T>>,
 {
-    fn intersects(&self, rhs: &G) -> bool {
+    fn intersects(&self, rhs: &Rect<T>) -> bool {
+        if self.is_empty() || rhs.is_empty() {
+            return false;
+        }
         if has_disjoint_bboxes(self, rhs) {
             return false;
         }
-        self.iter().any(|p| p.intersects(rhs))
+        // if no lines intersections, then linestring is either disjoint or within the polygon
+        self.0[0].intersects(rhs)
+            || self
+                .lines()
+                .any(|l| rhs.lines_iter().any(|other| l.intersects(&other)))
     }
 }
+
+impl<T> Intersects<Triangle<T>> for LineString<T>
+where
+    T: GeoNum,
+    Line<T>: Intersects<Line<T>>,
+    Coord<T>: Intersects<Triangle<T>>,
+{
+    fn intersects(&self, rhs: &Triangle<T>) -> bool {
+        if self.is_empty() || rhs.is_empty() {
+            return false;
+        }
+        if has_disjoint_bboxes(self, rhs) {
+            return false;
+        }
+        // if no lines intersections, then linestring is either disjoint or within the polygon
+        self.0[0].intersects(rhs)
+            || self
+                .lines()
+                .any(|l| rhs.lines_iter().any(|other| l.intersects(&other)))
+    }
+}
+
+//
+// MultiLineString Implementations
+//
+
+macro_rules! blanket_intersects_multilinestring {
```

Comment:
Unlike the LineString case, it doesn't look like there's any specialization going on here for MultiLineString. So can we keep the previous blanket implementation as it was?